core/src/main/scala/org/apache/spark/status/api/v1/OneStageResource.scala

```diff
@@ -0,0 +1,192 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.status.api.v1
+
+import javax.ws.rs._
+import javax.ws.rs.core.MediaType
+
+import org.apache.spark.SparkException
+import org.apache.spark.scheduler.StageInfo
+import org.apache.spark.ui.jobs.JobProgressListener
+import org.apache.spark.ui.jobs.UIData.StageUIData
+import org.apache.spark.util.SparkEnum
+
+@Produces(Array(MediaType.APPLICATION_JSON))
+private[v1] class OneStageResource(uiRoot: UIRoot) {
+
+  @GET
+  @Path("")
+  def stageData(
+    @PathParam("appId") appId: String,
+    @PathParam("stageId") stageId: Int
+  ): Seq[StageData] = {
+    forStage(appId, stageId){ (listener,stageAttempts) =>
+      stageAttempts.map { case (status, stageInfo) =>
+        val stageUiData = listener.synchronized {
+          listener.stageIdToData.get((stageInfo.stageId, stageInfo.attemptId)).
+            getOrElse(throw new SparkException("failed to get full stage data for stage: " +
+            stageInfo.stageId + ":" + stageInfo.attemptId)
+            )
+        }
+        AllStagesResource.stageUiToStageData(status, stageInfo, stageUiData,
+          includeDetails = true)
+      }
+    }
+  }
+
+  @GET
+  @Path("/{attemptId: \\d+}")
+  def oneAttemptData(
+    @PathParam("appId") appId: String,
+    @PathParam("stageId") stageId: Int,
+    @PathParam("attemptId") attemptId: Int
+  ): StageData = {
+    forStageAttempt(appId, stageId, attemptId) { case (status, stageInfo, stageUiData) =>
+      AllStagesResource.stageUiToStageData(status, stageInfo, stageUiData,
+        includeDetails = true)
+    }
+  }
+
+  @GET
+  @Path("/{attemptId: \\d+}/taskSummary")
+  def stageData(
+    @PathParam("appId") appId: String,
+    @PathParam("stageId") stageId: Int,
+    @PathParam("attemptId") attemptId: Int,
+    @DefaultValue("0.05,0.25,0.5,0.75,0.95") @QueryParam("quantiles") quantileString: String
+  ): TaskMetricDistributions = {
+    forStageAttempt(appId, stageId, attemptId) { case (status, stageInfo, stageUiData) =>
+      val quantiles = quantileString.split(",").map{s =>
+        try {
+          s.toDouble
+        } catch {
+          case nfe: NumberFormatException =>
+            throw new BadParameterException("quantiles", "double", s)
+        }
+      }
+      AllStagesResource.taskMetricDistributions(stageUiData.taskData.values, quantiles)
+    }
+  }
+
+  @GET
+  @Path("/{attemptId: \\d+}/taskList")
+  def taskList(
+    @PathParam("appId") appId: String,
+    @PathParam("stageId") stageId: Int,
+    @PathParam("attemptId") attemptId: Int,
+    @DefaultValue("0") @QueryParam("offset") offset: Int,
+    @DefaultValue("20") @QueryParam("length") length: Int,
+    @DefaultValue("ID") @QueryParam("sortBy") sortBy: TaskSorting
+  ): Seq[TaskData] = {
+    forStageAttempt(appId, stageId, attemptId) { case (status, stageInfo, stageUiData) =>
```

Comment:
Hmm... this threw me off a little bit. `forStageAttempt` sounds too much like `foreachStageAttemp`, while it's more like `withStageAttempt`. Meaning, the closure will just be executor on the single stage attempt being defined by the arguments, from what I understand. Maybe rename so it's clearer?